sql: allow " as an escape character for COPY

otan opened the pull request:
PostgreSQL allows `\\\"` as an escape for `\"`. However, we seem to not
allow it at all. This was a blocker for osm2pgsql.

In general, I think we should be allowing any non-captured escape to be
used as is, as PostgreSQL does:
```
otan=# \copy copytest from stdin
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> \1\2
>> COPY 1
otan=# select * from copytest
;
    t
----------
 \"
 \x01\x02
(2 rows)

otan=# \copy copytest from stdin
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> \a
>> COPY 1
otan=# select * from copytest
;
    t
----------
 \"
 \x01\x02
 a
(3 rows)

otan=# \copy copytest from stdin
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> \b
>> COPY 1
otan=# select * from copytest
;
    t
----------
 \"
 \x01\x02
 a
 \x08
(4 rows)
```

If you think we should change our code to follow this, let me know.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52044)
<!-- Reviewable:end -->

otan committed:
sql: allow " as an escape character for COPY

PostgreSQL allows `\\\"` as an escape for `\"`. However, we seem to not
allow it at all. This was a blocker for osm2pgsql.

In general, I think we should be allowing any non-captured escape to be
used as is, as PostgreSQL does:
```
otan=# \copy copytest from stdin
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> \1\2
>> COPY 1
otan=# select * from copytest
;
    t
----------
 \"
 \x01\x02
(2 rows)

otan=# \copy copytest from stdin
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> \a
>> COPY 1
otan=# select * from copytest
;
    t
----------
 \"
 \x01\x02
 a
(3 rows)

otan=# \copy copytest from stdin
Enter data to be copied followed by a newline.
End with a backslash and a period on a line by itself, or an EOF signal.
>> \b
>> COPY 1
otan=# select * from copytest
;
    t
----------
 \"
 \x01\x02
 a
 \x08
(4 rows)
```

If you think we should change our code to follow this, let me know.

Release note: None

mjibson commented:
From https://www.postgresql.org/docs/10/sql-copy.html:

> Any other backslashed character that is not mentioned in the above table will be taken to represent itself.

This is the actual problem. Even though your PR would fix the `\"` case, it wouldn't fix the `\a` case you also have in your terminal example above. The correct fix would be to do the above sentence (and the rest of that paragraph which deals with priority of special escaped sequences).

otan commented:
sg!

mjibson commented:
I'm working on COPY on my Fridays so I'll fix this.

otan commented:
well, saw this too late :P
